### PR TITLE
core(runner): independent gather and audit functions

### PR DESCRIPTION
**Pull request by adamraine**
Step 2 of #13364

Splits `Runner.run` into two functions, `Runner.gatherPhase` and `Runner.auditPhase`. Right now, these two functions are always used together, but step 3 of #13364 will start using these functions separately in use flows.

Ref
#11313

### CODE REVIEW

File: lighthouse-core/fraggle-rock/gather/navigation-runner.js [hunk 1]
```diff
@@ -287,7 +287,7 @@ async function navigation(options) {
     skipAboutBlank: configContext.skipAboutBlank,
   };
 
-  return Runner.run(
+  const gatherResult = await Runner.gatherPhase(
```
Comment (adamraine) [inline lighthouse-core/fraggle-rock/gather/navigation-runner.js:290]: Another idea I had was to store artifacts, config, cache, etc on an instance of `Runner` and pass that instance around instead of this new `gatherResult` object.

Comment (adamraine) [reply lighthouse-core/fraggle-rock/gather/navigation-runner.js]: Ya I like this better, gonna see what it looks like.

File: lighthouse-core/fraggle-rock/gather/navigation-runner.js [hunk 2]
```diff
@@ -287,7 +287,8 @@ async function navigation(options) {
     skipAboutBlank: configContext.skipAboutBlank,
   };
 
-  return Runner.run(
+  const runner = new Runner(config, computedCache);
```
Comment (brendankenny) [inline lighthouse-core/fraggle-rock/gather/navigation-runner.js:290]: IMO if it's just a glorified options object it should just be an options object. There's no complicated state being shared or encapsulated, it's just a cache and an immutable config, so no reason to have a constructed object.

Ideally these functions would be as independent as possible (possibly eventually in separate files) to avoid ease of unseparating concerns in the future

File: lighthouse-core/fraggle-rock/gather/navigation-runner.js [hunk 3]
```diff
@@ -287,7 +287,8 @@ async function navigation(options) {
...
+  await runner.gatherPhase(
```
Comment (brendankenny) [inline lighthouse-core/fraggle-rock/gather/navigation-runner.js:291]: We already use phase for the old before/afterPass and the new start/stopInstrumentation, so ideally we wouldn't overload it. gather and audit are already verbs, so the function names could be just those :)

**Review by brendankenny — COMMENTED** [review]
excited for this, though less so for this `new Runner()` interface :)

I was a fan of the
```js
/**
 * @template {LH.Config.Config | LH.Config.FRConfig} TConfig
 * @param {(runnerData: {config: TConfig, driverMock?: Driver}) => Promise<LH.Artifacts>} gatherFn
 * @param {{config: TConfig, driverMock?: Driver}} options
 * @return {Promise<LH.Artifacts>}
 */
static async gather(gatherFn, options) {}

/**
 * @template {LH.Config.Config | LH.Config.FRConfig} TConfig
 * @param {LH.Artifacts} artifacts
 * @param {{config: TConfig, computedCache: Map<string, ArbitraryEqualityMap>}} options
 * @return {Promise<LH.RunnerResult|undefined>}
 */
static async audit(artifacts, options) {}
```
approach

File: lighthouse-core/fraggle-rock/gather/navigation-runner.js (hunk 2 above)
Comment (adamraine) [reply lighthouse-core/fraggle-rock/gather/navigation-runner.js:290]: Aight I'll back down from this

File: lighthouse-core/fraggle-rock/gather/navigation-runner.js (hunk 3 above)
Comment (adamraine) [reply lighthouse-core/fraggle-rock/gather/navigation-runner.js:291]: Name change SGTM

File: lighthouse-core/runner.js [hunk 4]
```diff
@@ -212,8 +229,11 @@ class Runner {
         entryType: entry.entryType,
       };
     }).sort((a, b) => a.startTime - b.startTime);
-    const runnerEntry = timingEntries.find(e => e.name === 'lh:runner:run');
-    return {entries: timingEntries, total: runnerEntry?.duration || 0};
+    const gatherEntry = timingEntries.find(e => e.name === 'lh:runner:gather');
+    const auditEntry = timingEntries.find(e => e.name === 'lh:runner:audit');
+    const gatherTiming = gatherEntry?.duration || 0;
+    const auditTiming = auditEntry?.duration || 0;
+    return {entries: timingEntries, total: gatherTiming + auditTiming};
```
Comment (adamraine) [inline lighthouse-core/runner.js:236]: This might encounter issues in a flow with multiple steps. I don't know if a flow step is guaranteed to query the correct timing entry from the logger.

@brendankenny this might be a reason to use a `Runner` instance instead. We could store timing information on that instance instead of querying the logger.

File: lighthouse-core/runner.js [hunk 5]
```diff
@@ -44,24 +45,15 @@ class Runner {
...
-        category: 'lifecycle',
-        data: sentryContext?.extra,
-      });
-
-      const artifacts = await this.gatherAndManageArtifacts(gatherFn, runOpts);
-
       // Potentially quit early
       if (settings.gatherMode && !settings.auditMode) return;
 
       // Audit phase
```
Comment (connorjclark) [inline lighthouse-core/runner.js:51]: nit: remove comment

File: lighthouse-core/runner.js [hunk 6]
```diff
@@ -150,38 +137,68 @@ class Runner {
...
    * @param {(runnerData: {config: TConfig, driverMock?: Driver}) => Promise<LH.Artifacts>} gatherFn
-   * @param {{config: TConfig, driverMock?: Driver}} options
+   * @param {{config: TConfig, driverMock?: Driver, computedCache: Map<string, ArbitraryEqualityMap>}} options
    * @return {Promise<LH.Artifacts>}
    */
-  static async gatherAndManageArtifacts(gatherFn, options) {
+  static async gather(gatherFn, options) {
     const settings = options.config.settings;
 
     // Gather phase
```
Comment (connorjclark) [inline lighthouse-core/runner.js:146]: nit: remove comment (Gather Phase)

File: lighthouse-core/runner.js [hunk 7]
```diff
@@ -150,38 +137,68 @@ class Runner {
...
+  /**
+   * @param {any} err
+   * @param {LH.Config.Settings} settings
+   */
+  static async createRunnerError(err, settings) {
+    // i18n LighthouseError strings.
+    if (err.friendlyMessage) {
+      err.friendlyMessage = format.getFormatted(err.friendlyMessage, settings.locale);
+    }
+    await Sentry.captureException(err, {level: 'fatal'});
```
Comment (connorjclark) [inline lighthouse-core/runner.js:200]: this function could be sync, and we can do `Sentry.captrueException(...).catch(() => console.log('failed to send error to sentry'))` (or just do nothing).

We really don't want to have a fatal error just because `Sentry.captureException` decided to throw.

File: lighthouse-core/test/runner-test.js [hunk 8]
```diff
@@ -49,6 +49,13 @@ describe('Runner', () => {
   /** @type {jest.Mock} */
   let runAuditSpy;
 
+  beforeAll(() => {
+    Runner.run = async (gatherFn, opts) => {
```
Comment (connorjclark) [inline lighthouse-core/test/runner-test.js:53]: how about a new function in this file, instead of augmenting the module?

**Review by connorjclark — CHANGES_REQUESTED** [review]

**Review by connorjclark — APPROVED** [review]

File: lighthouse-core/runner.js (hunk 4 above)
Comment (adamraine) [reply lighthouse-core/runner.js:235]: It's not causing any issues right now because `log.takeTimingEntries` clears everything between each flow step. I'll investigate this in the next part. There's probably a way to remember timings for each step without adding state to `Runner`.

File: clients/test/lightrider-entry-test.js [hunk 9]
```diff
@@ -59,7 +59,7 @@ describe('lightrider-entry', () => {
     });
 
     it('specifies the channel as lr', async () => {
-      const runStub = jest.spyOn(Runner, 'run');
+      const runStub = jest.spyOn(Runner, 'gather');
```
Comment (brendankenny) [inline clients/test/lightrider-entry-test.js:62]: not that these spy tests are the best, but seems like these should be on the higher-level `lighthouse()` function now

File: lighthouse-core/test/runner-test.js [hunk 10]
```diff
@@ -38,6 +38,11 @@ describe('Runner', () => {
     };
   };
 
+  const runGatherAndAudit = async (gatherFn, opts) => {
```
Comment (brendankenny) [inline lighthouse-core/test/runner-test.js:41]: it's definitely a bit weird to refactor the runner interface but then not have any changes in its unit tests :) These are kind of `lighthouse-core/index.js` tests now instead.

- what are the new use cases being opened up by exposing both `gather` and `audit` in Runner? #13364 basically
- are those cases covered by tests?

File: clients/test/lightrider-entry-test.js (hunk 9 above)
Comment (adamraine) [reply clients/test/lightrider-entry-test.js:62]: One of the tests verifies that the `config.settings.channel` is set to `lr` before `config` reaches `Runner`. I think there is still value in testing `Runner.gather` rather than `lighthouse()`.

File: lighthouse-core/test/runner-test.js (hunk 10 above)
Comment (adamraine) [reply lighthouse-core/test/runner-test.js:41]: Ya we could have a test that does multiple calls to `gather` followed by multiple corresponding calls to `audit`

File: lighthouse-core/runner.js (hunk 4 above)
Comment (brendankenny) [reply lighthouse-core/runner.js:239]: FWIW it's not clear what `total` should mean going forward. Right now it's the total of the LH run that _just_ completed, so if run with `-A`, the `-G` timings are in the `entries` array, but not included in the total. That would be changed here, and I kind of prefer the old way, but I'm not sure if the `total` timing of `-A` runs is important enough to care (if e.g. you're trying to optimizing something, you're likely looking at some other timing or tool).

In a user flow, is the amount of time to gather and audit a particular step the most meaningful interpretation of "total"? Seems ok.

Should add some total tests if we're making a decision.

Also it's worth pointing out that `total` was a placeholder while we figured out `entries` and only kept it around because it was kind of useful after `entries` landed, so the only bar to clear here is "kind of useful" :)

File: lighthouse-core/runner.js [hunk 11]
```diff
@@ -150,38 +137,71 @@ class Runner {
...
+          const path = this._getDataSavePath(settings);
+          await assetSaver.saveArtifacts(artifacts, path);
+        }
       }
+
+      log.timeEnd(runnerStatus);
+
+      // If `gather` is run multiple times before `audit`, the timing entries for each `gather` can pollute one another.
+      // Timing entries are stored in artifacts.Timing, so we can clear the timing entries here.
+      log.takeTimeEntries();
```
[the remark this answers is not in the document]
Comment (adamraine) [reply lighthouse-core/runner.js:186]: > If there aren't going to be any measures that go over the entire run, then this works, but then the base artifact finalization should be able to call it directly.

I tried that at first, it clears out the marker used by the `log.timeEnd` right above this. Had to clear the timings after it.

File: lighthouse-core/runner.js [hunk 12]
```diff
@@ -134,12 +126,7 @@ class Runner {
 
       return {lhr, artifacts, report};
     } catch (err) {
-      // i18n LighthouseError strings.
-      if (err.friendlyMessage) {
-        err.friendlyMessage = format.getFormatted(err.friendlyMessage, settings.locale);
-      }
-      await Sentry.captureException(err, {level: 'fatal'});
-      throw err;
+      throw Runner.createRunnerError(err, settings);
```
[the remark this answers is not in the document]
Comment (adamraine) [reply lighthouse-core/runner.js:129]: I can't find a place where a "friendly" error makes it all the way up here. They should all be caught be the catch block in `_runAudit`.

File: lighthouse-core/runner.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (brendankenny) [reply lighthouse-core/runner.js:186]: isn't it cleared out regardless by this call since `log.takeTimeEntries()` clears them and we don't retain the returned value?

[the remark this answers is not in the document]
Comment (adamraine) [reply lighthouse-core/runner.js:186]: Sorry, it clears `lh:runner:gather` not `lh:runner:audit` before its `log.timeEnd` is called.

File: lighthouse-core/runner.js [hunk 13]
```diff
@@ -150,38 +137,67 @@ class Runner {
...
-      artifacts = await gatherFn({
-        config: options.config,
-        driverMock: options.driverMock,
+    try {
+      const runnerStatus = {msg: 'Gather phase', id: 'lh:runner:gather'};
+      log.time(runnerStatus, 'verbose');
+
+      const sentryContext = Sentry.getContext();
+      Sentry.captureBreadcrumb({
+        message: 'Run started',
```
[the remark this answers is not in the document]
Comment (adamraine) [reply lighthouse-core/runner.js:153]: FWIW the FR runners should call `Runner.gather` every time, but there is no guarantee that the FR runners themselves will be run. The only question is what happens in `user-flow.js`. Should we consider each flow step an independent "run" from the sentry perspective, or should the entire flow be a single "run".

[the remark this answers is not in the document]
Comment (brendankenny) [reply lighthouse-core/runner.js:153]: I thought from #13364 that they wouldn't be calling into gather() anymore, since it's kind of redundant (they already define their own gather function)? With multiple gathers, at least, -G and -A will have to be tweaked to work.

File: lighthouse-core/runner.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (brendankenny) [reply lighthouse-core/runner.js:186]: That's what I meant as well, I think. That's the change in sample_v2 in the CI basics failure?

sorry you had to step into the mess that is collecting our timings inside the thing being timed for this PR (ask @connorjclark how the last time he made a change to them went...).

[the remark this answers is not in the document]
Comment (adamraine) [reply lighthouse-core/runner.js:186]: Yeah I'm currently trying to unravel the CI failure. Long story short I don't think this is the correct approach anymore :)

File: lighthouse-core/runner.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (adamraine) [reply lighthouse-core/runner.js:153]: > I thought from #13364 that they wouldn't be calling into gather() anymore

Using `Runner.gather()` in the FR runners is still useful because it gives us the timing entry, error handling, and the sentry call. As you said, we could add those things to each FR runner.

I did update #13364 to reflect that but didn't announce the change sorry.

> With multiple gathers, at least, -G and -A will have to be tweaked to work.

This is true, but we don't *need* to get rid of the logic entirely if the other stuff in `gather()` is still useful. The issue should only be a problem for user flows which will have it's own save/load artifacts system.

**Review by paulirish — COMMENTED** [review]
looks nice

File: lighthouse-cli/test/smokehouse/report-assert.js [hunk 14]
```diff
@@ -111,6 +111,35 @@ function findDifference(path, actual, expected) {
     const keyPath = path + keyAccessor;
     const expectedValue = expected[key];
 
+    if (key === '_includes') {
```
Comment (connorjclark) [inline lighthouse-cli/test/smokehouse/report-assert.js:114]: This needs an README update. and should probably be a separate PR.

Comment (adamraine) [reply lighthouse-cli/test/smokehouse/report-assert.js:114]: It was a separate PR, are you saying you want to merge https://github.com/GoogleChrome/lighthouse/pull/13614 separately?
